### PR TITLE
DAGE-8: Abstract Writter + Quepid + Testing

### PR DESCRIPTION
**Pull request by Intrinsical-AI**
Implement AbstractWriter as a base class for data exporters.
Implement QuepidWriter to export data to a CSV format compatible with Quepid.
Add comprehensive unit tests for the writers, including edge cases.

### CODE REVIEW

File: rre-dataset-generator/src/writters.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import csv
+from typing import List, Tuple
+from abc import ABC, abstractmethod
+from src.search_engine.data_store import DataStore
+from src.model.query_rating_context import QueryRatingContext
+
+class AbstractWriter(ABC):
```
Comment (nseidan) [inline rre-dataset-generator/src/writters.py]: the class name `AbstractWriter` is very generic, better to change to meaningful name, maybe DataStoreWriter/DataStoreExporter etc.

File: rre-dataset-generator/src/writters.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    QuepidWriter: Write the data structure to a Quepid format (CSV).
+    The format is: query,docid,rating
+    """
+    def write(self, output_path: str) -> None:
+        """
+        Writes queries and their scored documents to a CSV file in Quepid format.
+        """
+        with open(output_path, 'w', newline='') as csvfile:
+            writer = csv.writer(csvfile)
+            writer.writerow(['query', 'docid', 'rating'])
```
Comment (nseidan) [inline rre-dataset-generator/src/writters.py]: proposal: `doc_id, rating_score`

File: rre-dataset-generator/src/writters.py (hunk 1 above)
Comment (nseidan) [reply rre-dataset-generator/src/writters.py]: By the way, the file name has a typo `writters.py`, extra  't' . Actually I'd suggest to have a better name, writers means anything. 
Also I would split the file into two files: BaseClassWriter (aka AbstractWriter) in `base_class_writer.py `and QuepidWriter in  `quepid_writer.py`.

File: rre-dataset-generator/tests/unit/test_writters.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
+import pytest
+import csv
+from pathlib import Path
+
+from src.search_engine.data_store import DataStore
```
Comment (nseidan) [inline rre-dataset-generator/tests/unit/test_writters.py]: file name has a typo -> `test_writters.py`, extra 't'

**Review by nicolo-rinaldi — APPROVED** [review]

File: rre-dataset-generator/tests/unit/test_writters.py (hunk 3 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/tests/unit/test_writters.py]: Right, there's everywhere in this commit

File: rre-dataset-generator/src/writters.py (hunk 2 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/src/writters.py]: doc_id makes sense instead docid for redability
rating_score seems too verbose / redundant for me for few fields

File: rre-dataset-generator/src/writters.py (hunk 1 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/src/writters.py]: The name seems fine for me tbh. The split it's too time / space consuming for me for only 2 real writters (REE, QUEPID) Danield told to write an independant file when I asked 

https://sease.slack.com/archives/C093CLEDZLJ/p1752748942553739?thread_ts=1752748798.430339&cid=C093CLEDZLJ

File: rre-dataset-generator/src/writters.py (hunk 2 above)
Comment (nseidan) [inline rre-dataset-generator/src/writters.py]: also we might need to log (debug mode ? ) while QuepidWriter writes into a file.

**Review by nseidan — APPROVED** [review]

File: rre-dataset-generator/src/writters.py
Comment (dantuzi) [inline rre-dataset-generator/src/writters.py:1]: Can we have a single file for the AbstractWriter and another file for the Quepid writer?
PS: There is a typo in the writer's name (one 't').

**Comment by nseidan** [comment]
@Intrinsical-AI I approved but please take into account  my suggestions at least typos, docid, logging.

File: rre-dataset-generator/src/writters.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+class QuepidWriter(AbstractWriter):
+    """
+    QuepidWriter: Write the data structure to a Quepid format (CSV).
+    The format is: query,docid,rating
+    """
+    def write(self, output_path: str) -> None:
+        """
+        Writes queries and their scored documents to a CSV file in Quepid format.
+        """
+        with open(output_path, 'w', newline='') as csvfile:
```
Comment (dantuzi) [inline rre-dataset-generator/src/writters.py]: How about if the output file doesn't exist? Does the method `open(output_path, 'w', newline='')` create the output file?

File: rre-dataset-generator/tests/unit/test_writters.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+from src.writters import QuepidWriter, AbstractWriter
+
+
+@pytest.fixture
```
Comment (dantuzi) [inline rre-dataset-generator/tests/unit/test_writters.py]: what does this annotation do?

File: rre-dataset-generator/tests/unit/test_writters.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+def unrated_datastore() -> DataStore:
+    """Returns a DataStore with unrated documents."""
+    datastore = DataStore()
+    datastore.add_query("query 1", "doc1")
+    datastore.add_query("query 2", "doc2")
+    return datastore
+
+
+class TestAbstractWriter:
+    def test_get_queries_with_ratings(self, populated_datastore):
```
Comment (dantuzi) [inline rre-dataset-generator/tests/unit/test_writters.py]: It's probably too long name, but I'd like to have something like
```
def test_get_queries_with_ratings_expect_all_is_retrieved
```

File: rre-dataset-generator/tests/unit/test_writters.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            ("test query 1", "doc1", 1),
+            ("test query 1", "doc2", 2),
+            ("test query 2", "doc4", 3),
+        }
+
+        assert set(rated_items) == expected_items
+
+
+class TestQuepidWriter:
+    def test_write(self, populated_datastore, tmp_path: Path):
```
Comment (dantuzi) [inline rre-dataset-generator/tests/unit/test_writters.py]: test_write_expect_file_successfully_written

File: rre-dataset-generator/tests/unit/test_writters.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+            rows = set(tuple(row) for row in reader)
+            expected_rows = {
+                ('test query 1', 'doc1', '1'),
+                ('test query 1', 'doc2', '2'),
+                ('test query 2', 'doc4', '3'),
+            }
+            assert rows == expected_rows
+
+    def test_write_with_empty_datastore(self, empty_datastore, tmp_path: Path):
```
Comment (dantuzi) [inline rre-dataset-generator/tests/unit/test_writters.py]: test_write_with_empty_datastore_expect_outpufile_written_with_only_header

File: rre-dataset-generator/tests/unit/test_writters.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+        with open(output_file, 'r', newline='') as csvfile:
+            reader = csv.reader(csvfile)
+            header = next(reader)
+            assert header == ['query', 'docid', 'rating']
+            rows = list(reader)
+            assert len(rows) == 1
+            assert rows[0] == [query_text, doc_id, '1']
+
+    def test_write_with_zero_rating(self, tmp_path: Path):
```
Comment (dantuzi) [inline rre-dataset-generator/tests/unit/test_writters.py]: What does it test more than the happy case at line 47?

File: rre-dataset-generator/tests/unit/test_writters.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+        output_file = tmp_path / "output.csv"
+        writer = QuepidWriter(unrated_datastore)
+        writer.write(str(output_file))
+
+        with open(output_file, 'r', newline='') as csvfile:
+            reader = csv.reader(csvfile)
+            header = next(reader)
+            assert header == ['query', 'docid', 'rating']
+            # Check that there are no more rows
+            with pytest.raises(StopIteration):
```
Comment (dantuzi) [inline rre-dataset-generator/tests/unit/test_writters.py]: Why didn't you do as you did below? The following lines don't work?
```
rows = list(reader)
assert len(rows) == 0
```

File: rre-dataset-generator/src/writters.py (hunk 4 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/src/writters.py]: With w yes.

If the file doesn’t exist, it creates a new empty file at output_path.

If the file does exist, it overwrites it.

---

'a' it's for append

File: rre-dataset-generator/tests/unit/test_writters.py (hunk 5 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/tests/unit/test_writters.py]: it's a decorator for a function that sets up some state or resource (e.g., a variable, object) for using in following tests.

File: rre-dataset-generator/tests/unit/test_writters.py (hunk 6 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/tests/unit/test_writters.py]: I removed that test class in the last PR as the name was not aligned

File: rre-dataset-generator/tests/unit/test_writters.py (hunk 10 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/tests/unit/test_writters.py]: more human approach indeed

File: rre-dataset-generator/tests/unit/test_writters.py (hunk 5 above)
Comment (Intrinsical-AI) [reply rre-dataset-generator/tests/unit/test_writters.py]: The main difference here:
"@pytest.fixture
def populated_datastore() -> DataStore:"

I think it's that we're creating the instance of DataStore(), and also adding documents in a single step. While after we can call in other tests functions

"class TestQuepidWriter:
    def test_write_expect_file_successfully_written(self, populated_datastore,"
    
    
  And pytests will handle everything for us ["Any test function that includes populated_datastore as an argument will automatically receive that pre-set object, without needing to know how to create it."]

**Comment by Intrinsical-AI** [comment]
Corrected double 'tt' typos, removed and renamed tests.

File: rre-dataset-generator/src/writers.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                rating = query_ctx.get_rating_score(doc_id)
+                if rating != QueryRatingContext.DOC_NOT_RATED:
+                    result.append((query_text, doc_id, rating))
+        return result
+
+
+class QuepidWriter(AbstractWriter):
+    """
+    QuepidWriter: Write the data structure to a Quepid format (CSV).
+    The format is: query,docid,rating
```
Comment (nseidan) [inline rre-dataset-generator/src/writers.py:39]: @Intrinsical-AI forgot to update the comment for doc_id
